### PR TITLE
Use cases: elektra-web

### PR DESCRIPTION
**Pull request by omnidan**
# Purpose

Use cases for elektra-web.
# Checklist
- [X] commit messages are fine (with references to issues)

@markus2330 please review my pull request

### CODE REVIEW

File: doc/usecases/elektra_web/UC_add_instance.md [hunk 1]
```diff
@@ -0,0 +1,20 @@
+# Use Case: Add instance
```
Comment (markus2330) [inline doc/usecases/elektra_web/UC_add_instance.md]: Where will it be added?

File: doc/usecases/elektra_web/UC_add_instance.md [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+## Summary
+
+Title: Add instance
+Scope: Setup
+Level: User Goal
+Actors: User (usually a sysadmin)
+Brief: User sets up elektrad on the instance and connects it to clusterd according to the corresponding README.
```
Comment (markus2330) [inline doc/usecases/elektra_web/UC_add_instance.md]: Try to be a bit more high-level, do not mention elektrad or similar.

Actually "add instance" and "add cluster" should be merged to "describe network topology"

File: doc/usecases/elektra_web/UC_add_cluster.md [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+Title: Add cluster
+Scope: Setup
+Level: User Goal
+Actors: User (usually a sysadmin)
+Brief: User groups multiple instances into a cluster.
+
+## Scenarios
+
+Precondition: Add instance (multiple times)
+Main success scenario: User successfully creates a cluster of multiple instances.
```
Comment (markus2330) [inline doc/usecases/elektra_web/UC_add_cluster.md]: Recursion missing.

File: doc/usecases/elektra_web/UC_cluster_config.md [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+Scope: Configuration
+Level: User Goal
+Actors: User (usually a sysadmin)
+Brief: User configures a cluster (multiple instances at once) via the client.
+
+## Scenarios
+
+Precondition: Add cluster
+Main success scenario: User successfully changes configuration of the cluster.
+Alternative scenario: One of the instances is not online. The user is informed about the issue.
```
Comment (markus2330) [inline doc/usecases/elektra_web/UC_cluster_config.md]: Does this mean nothing happens here except a popup?

File: doc/usecases/elektra_web/UC_cluster_config.md [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+Error scenario: Technical problems while persisting configuration to one of the instances. The user is informed about the issue.
```
Comment (markus2330) [inline doc/usecases/elektra_web/UC_cluster_config.md]: Again: What happens, are there any changes anywhere?

You could describe the behaviour also in "Postcondition".

File: doc/usecases/elektra_web/UC_multiple_config.md [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+Scope: Configuration
+Level: User Goal
+Actors: User (usually a sysadmin)
+Brief: User configures multiple (standalone) instances via the client.
+
+## Scenarios
+
+Precondition: Add instance (multiple times)
+Main success scenario: User successfully changes configuration of multiple instances.
+Alternative scenario: Instance not online. The user is informed about the issue.
```
Comment (markus2330) [inline doc/usecases/elektra_web/UC_multiple_config.md]: Why is it possible to select offline nodes?

Better say: "A node got offline ..."

File: doc/usecases/elektra_web/UC_remove_cluster.md [hunk 7]
```diff
@@ -0,0 +1,19 @@
+# Use Case: Remove cluster
```
Comment (markus2330) [inline doc/usecases/elektra_web/UC_remove_cluster.md]: I thought the browser is directly connected to a clusterd? Maybe you need an alternative scenario if the user tries to do that?

File: doc/usecases/elektra_web/UC_remove_instance.md [hunk 8]
```diff
@@ -0,0 +1,19 @@
+# Use Case: Remove instance
+
+## Summary
+
+Title: Remove instance
+Scope: Setup
+Level: User Goal
+Actors: User (usually a sysadmin)
+Brief: User removes an instance via the client.
```
Comment (markus2330) [inline doc/usecases/elektra_web/UC_remove_instance.md]: Generally remove instance or remove instance from cluster?

File: doc/usecases/elektra_web/UC_single_config.md [hunk 9]
```diff
@@ -0,0 +1,19 @@
+# Use Case: Remote configuration of a single instance
```
Comment (markus2330) [inline doc/usecases/elektra_web/UC_single_config.md]: Is this the use case when browser is also connected to single instance?

File: doc/usecases/elektra_web/UC_add_cluster.md (hunk 3 above)
Comment (markus2330) [inline doc/usecases/elektra_web/UC_add_cluster.md]: They could be interesting to:

1. define multiple groups if you use a cluster per application (all mail, web servers ...)
2. for failover (two clusterd for endpoints refer to other clusters, if one machine is gone, the system is still reachable)

Also only connecting to elektrad is essential, nodejs is a too heavy dependency for many systems.

**Comment by Namoshek** [comment]
During review of the PR, I came across this:

>> omnidan:
>> the client actually doesn't access elektrad directly, clusterd acts as a kind of router that maintains instances (this makes it easy to have a client that can be accessed from various devices, otherwise I'd have to store the connection info for clusterd/elektrad instances in localstorage, cookies or something similar). this way all you have to provide in the client is the host of the clusterd instance.

> markus2330:
> Should also be discussed in use cases. If you also provide a lightweight implementation of clusterd, the dependency might be ok.

I think @omnidan's approach is good. If he opens `elektrad` servers to the web client directly, he also has to give authorization data to the web client (which is currently not the case - the client only holds authorization for the `clusterd` server). I also think that the hierarchical design allows for more security through firewalls. For example could the `clusterd` server be reachable for the public, but all other servers (`elektrad`) only through the intranet.

**Comment by omnidan** [comment]
- merged "add cluster" and "add instance" into "describe network topology"
- described adding instances/clusters in a more high-level way
- clarified what happens when an instance in a cluster is not available
- removed "multiple config" because it just meant executing "single config" on multiple instances one after another (not a separate use case)
- added "undo/redo" use case

> I thought the browser is directly connected to a clusterd? Maybe you need an alternative scenario if the user tries to do that?
clusterd actually manages all clusters, I don't think it makes much sense to have recursion of clusterd's, because there are many issues with that approach (see @Namoshek's comment) It would also make the setup and maintainance a lot harder.

> Generally remove instance or remove instance from cluster?
generally removing instances

> Is this the use case when browser is also connected to single instance?
no, this is also done via clusterd

> add keys, change keys, find keys, arrays
this is part of the "single config" and "cluster config" use cases, maybe I should make a new "configuration" use case that further describes these operations?

> import keys copy&paste
what do you mean? there will be an interactive tree view where you can create, delete and modify keys (this is the single/cluster config use case)

> help system
not sure what is meant by that, the UI should be intuitive, if something is unclear I will add short info blocks or tooltips, there shouldn't be a need for a separate help system

**Comment by markus2330** [comment]
> import keys copy&paste

I mean that existing keys can be copied around via shortcuts (ctrl+c, ctrl+v) or drag&drop.

> help system

I also mean help about the configuration itself. So metadata like "description" should be nicely displayed (e.g. in tool tips when hovering keys).

**Comment by omnidan** [comment]
> I mean that existing keys can be copied around via shortcuts (ctrl+c, ctrl+v) or drag&drop.

Key contents can already be copied around via shortcuts, or do you mean whole sets of keys? Drag & drop for moving keys around should be possible.

> I also mean help about the configuration itself. So metadata like "description" should be nicely displayed (e.g. in tool tips when hovering keys).

Ah, yes I was going to use metadata to adjust the UI (displaying number fields in a special way, validating regex constraints, adding descriptions as tooltips, etc)

---

What is the best way to document these features? Should I make a "configuration" use case that "single config" and "cluster config" both make use of? The same configuration interface would be used for configuring single instances and clusters.

**Comment by omnidan** [comment]
@markus2330 I updated the use cases, they should now describe the focus of elektra-web more (configuration through an interactive tree view).

**Comment by markus2330** [comment]
There are still two comments without response.

File: doc/usecases/elektra_web/UC_remove_cluster.md (hunk 7 above)
Comment (omnidan) [inline doc/usecases/elektra_web/UC_remove_cluster.md]: clusterd actually manages all clusters, I don't think it makes much sense to have recursion of clusterd's, because there are many issues with that approach (see @Namoshek's comment) It would also make the setup and maintainance a lot harder.

File: doc/usecases/elektra_web/UC_single_config.md (hunk 9 above)
Comment (omnidan) [inline doc/usecases/elektra_web/UC_single_config.md]: no, this is also done via clusterd

**Comment by omnidan** [comment]
@markus2330 I already responded to them in the comment above: https://github.com/ElektraInitiative/libelektra/pull/999#issuecomment-260984107 - but for a better overview I also pasted the answers into the comment threads on the diff. Let me know if anything is still unclear.

**Comment by markus2330** [comment]
I am still not convinced about problems with recursion. The main problem is if the same node is present multiple times, and this can happen regardless of recursion. Can you write a decision about why you decide to do the clustering the way you do and how to deal with conflicts and the same node added to multiple clusters? (Separate PR)

And I also would like to see the single node use case to be supported better. It is okay if two node instances need to be running, but nevertheless there should be a really easy way to set this up. There should be a dedicated use case for this thus I think it will be the first contact for everyone trying out your tool. "Setup instance" is really vague. Please describe what the user must do.

And please self-review the use cases once more to make them more clear. E.g. why is there remove cluster if we have a use case "describe network"? Many questions are unanswered: What can we describe? What does the user get from describing it? (Why would he/she want to do it? Most other use cases refer to "Add instance/cluster".) These are only examples, please check them thoroughly for consistency.

**Comment by omnidan** [comment]
@markus2330 I updated the use cases and added a decision for the structure (#1173)

**Comment by omnidan** [comment]
@markus2330 is there still anything missing here? I have created decisions for the (recursive) structure: #1173 #1354

**Comment by markus2330** [comment]
Thus you (understandably and well argued) use simple techniques for conflict resolution and disallow inserting the same node multiple times as well as recursion; and use nodejs as only language for implementation (also understandable); please add more focus for usability. For example, there is still no use case about validation. (It is only very vaguely mentioned).

So for example, when `check/validation` defines a regex, you should also define it as input validation, without round-trips to the backend. We could also check in a study with and without this feature if users experience differences.

**Comment by markus2330** [comment]
Maybe a clarification: Of course you should not write a use case to avoid round-trips (maybe the round-trip does not matter at all, then simply forget my implementation suggested). Instead the use cases should be something like "users tries to input an erroneous string; no character is added in the input field and user gets feedback..."

**Comment by omnidan** [comment]
@markus2330 yes, I agree that validation should be done on the client (e.g. it would mark the input field as red and not send it to the backend at all)

I added a use case explaining validation.

File: doc/usecases/elektra_web/UC_key_validation.md [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+
+Precondition: Modifying keys
+Main success scenario: User edits a key's contents in a field that is
+  specifically formatted according to its metadata (e.g. number/date fields vs
+  string fields with regex validation). If the entered data passes the
+  validation, it is sent to the backend and saved to the key database.
+Alternative scenario: User enters data that fails the validation, the input
+  field is marked as invalid (e.g. with a red color and some information text)
+  and the data is NOT sent to the backend.
+Error scenario: Technical problems while persisting to the key database. The
```
Comment (markus2330) [inline doc/usecases/elektra_web/UC_key_validation.md]: Is this a copy&paste problem? For me it is not clear when this use case applies. Isn't it before modifying keys and/or persisting keys? (So the user only typing into input fields without pressing ok).

Usually there is not only an invalid state but also a state where validation might pass if the user continues typing.

Comment (omnidan) [inline doc/usecases/elektra_web/UC_key_validation.md]: in a way it's part of the "modifying keys" use case. my idea was that it will keep validating and persist only when the input becomes valid.

Comment (markus2330) [inline doc/usecases/elektra_web/UC_key_validation.md]: Yes, the idea is good, but this is not whats written in this use case. In the use case you have "modifying keys" as precondition. Actually it is the other way round: modification needs the key validation as precondition.

Ideally you think through the steps for a user again and compare it carefully with what currently is written.

Comment (omnidan) [inline doc/usecases/elektra_web/UC_key_validation.md]: @markus2330 I swapped out the preconditions now.

Comment (markus2330) [inline doc/usecases/elektra_web/UC_key_validation.md]: The issue is not only there. A general line in what the user can do when seems to be missing. For example, why is there no validation when "Adding keys"? Maybe you need a Use Case Diagram or some overview description to get this straight.

Please rethink the whole use cases from a users perspective, rebase, and let us finally merge this PR!

Comment (omnidan) [inline doc/usecases/elektra_web/UC_key_validation.md]: @markus2330 I think the issue is that validation is supposed to be everywhere, it's not a separate use case, it's integrated in all other use cases.

Comment (omnidan) [inline doc/usecases/elektra_web/UC_key_validation.md]: So maybe the best way to go about this is removing the validation use case and adapting all other use cases to describe the validation there?

Comment (markus2330) [inline doc/usecases/elektra_web/UC_key_validation.md]: No, use cases will always somehow interrelate, you cannot copy everything (you copied too much, hence the inconsistency issues now). For example undo/redo also depends on most other use cases. You should rethink how they depend on each other and fix currently wrong precondition. Maybe a behavioral UML diagram helps you to do so. Or read a bit more about how to create use cases.

Comment (omnidan) [inline doc/usecases/elektra_web/UC_key_validation.md]: Maybe I'm misunderstanding something, but while undo/redo is definitely its own use case (it's a separate feature that is triggered with certain buttons), validation is a feature that is integrated in all use cases that include text fields.

That's also what makes it hard to have a definitive order of preconditions - because validation happens until the field is valid, in which case it is stored on the server.

If I was gonna make it as a separate use case, I would have to copy & paste the same validation use case for various text fields, with little differences. So the best way to go about this would be explaining validation in the modifying keys use case, but also in the adding keys use case, and others.

Comment (markus2330) [inline doc/usecases/elektra_web/UC_key_validation.md]: Maybe you are trying to say too much with the use cases? I only noticed several times that it is very easy to find inconsistencies and points where clarifications are needed.

For example, now that you say it: undo is supported after every other *use case* (e.g. finding keys)? (Rather it only applies to actions that changed the state of the system.)

In the same way the validation only applies when the user edits something in the tree view. It shouldn't be too hard to say that explicitly.

Basically I am okay with every description that:

1. clearly describes what the user can achieve with your proposed system
2. are internally consistent

Comment (omnidan) [inline doc/usecases/elektra_web/UC_key_validation.md]: So you are proposing to make a generic validation use case with multiple use cases listed as precondition? but in that case, what would be the postcondition? would it be "continue with the post condition of the use case"? that's what confuses me, validation is neither before or after editing a field, it's something inbetween/an extension of use cases.

Comment (markus2330) [inline doc/usecases/elektra_web/UC_key_validation.md]: Extension (and inclusion) is something that is quite well defined in UML use case diagram, thus I suggested you could use that.

There is more than one correct way to describe it. It would be better if you can stick to the standard, it should have more than enough possibilities to say what you want.

You do not have to be complete. But it should be clear and consistent.

Comment (markus2330) [inline doc/usecases/elektra_web/UC_key_validation.md]: (You can also use UML features in the textual variant, if you do not want to draw something.)

Comment (omnidan) [inline doc/usecases/elektra_web/UC_key_validation.md]: Thanks for the hints, I will look into the UML standard more and then rework all use cases accordingly.

Comment (markus2330) [inline doc/usecases/elektra_web/UC_key_validation.md]: Thanks! It shouldn't take long, but you cannot fix it without going through the use cases completely.

**Comment by markus2330** [comment]
Please remove "work in progress" when updated.

**Comment by omnidan** [comment]
@markus2330 I reworked all the use-cases and used "extends" for key validation. I also removed everything related to clusters, it is still possible to have multiple instances, just not in clusters. I also split up some use-cases in the configuration tree view, as this is the focus of the application (not cluster configuration).

**Comment by omnidan** [comment]
The path through the use cases is as follows:

- Setup
- Setup instance
- View config

Now we are in the main part of the application, and we can do the following use cases:

- Adding keys
- Modifying keys
- Duplicating keys
- Drag & Drop keys
- Finding keys

Any of these use cases (except for finding because no change is made), are extended by the *Key validation* use case. After doing any of these use cases, we can do the *Undo/Redo* use case to revert them.

File: doc/usecases/elektra_web/UC_setup.md [hunk 11]
```diff
@@ -0,0 +1,21 @@
+# Use Case: Setup
```
Comment (markus2330) [inline doc/usecases/elektra_web/UC_setup.md]: Please write a README.md containing the info of the path trough the use cases.

File: doc/usecases/elektra_web/UC_undo_redo.md [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+Level: User Goal
+Actors: User (usually a sysadmin)
+Brief: User undos/redos changes to a configuration.
+
+## Scenarios
+
+Precondition: Modifying keys, Adding keys, Duplicating keys, Drag & Drop keys
+Main success scenario: User successfully undos/redos a configuration change.
+Alternative scenario: Instance not online. The user is informed about the issue.
+  If a cluster, changes will be written when the affected instance comes back
```
Comment (markus2330) [inline doc/usecases/elektra_web/UC_undo_redo.md]: Still clusters?

File: doc/usecases/elektra_web/UC_undo_redo.md [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+	online.
+Error scenario: Technical problems while configuring the instance. The user is
+  informed about the issue.
+Postcondition: The updated configuration is persisted to the instance/s.
+Non-functional Constraints:
+	- Ideally every operation should be undoable
```
Comment (markus2330) [inline doc/usecases/elektra_web/UC_undo_redo.md]: Ideally?

File: doc/usecases/elektra_web/UC_duplicating_keys.md [hunk 14]
```diff
@@ -0,0 +1,20 @@
+# Use Case: Duplicating keys
```
Comment (markus2330) [inline doc/usecases/elektra_web/UC_duplicating_keys.md]: Do we really need this as extra usecase? Isn't this copy&paste or ctrl+drag&drop?

File: doc/usecases/elektra_web/UC_duplicating_keys.md [hunk 15]
```diff
@@ -0,0 +1,20 @@
...
+
+## Scenarios
+
+Precondition: View configuration of an instance
+Main success scenario: User duplicates an existing key (and its subtree).
+Alternative scenario: User enters data that violates a validation rule, an error
+  message is shown.
+Error scenario: Technical problems while persisting to the key database. The
+  user is informed about the issue.
+Postcondition: The new key is persisted to the database.
```
Comment (markus2330) [inline doc/usecases/elektra_web/UC_duplicating_keys.md]: new key (and its subtree)

File: doc/usecases/elektra_web/UC_finding_keys.md [hunk 16]
```diff
@@ -0,0 +1,20 @@
...
+Scope: Configuration
+Level: User Goal
+Actors: User (usually a sysadmin)
+Brief: User searches for a key by filtering the key database.
+
+## Scenarios
+
+Precondition: View configuration of an instance
+Main success scenario: User searches for a key by entering information (its path
+  or data) in a field. The whole key database is filtered according to the
```
Comment (markus2330) [inline doc/usecases/elektra_web/UC_finding_keys.md]: data -> value.

Search for key is much faster, what about first show by key and continue searching for values in background?

File: doc/usecases/elektra_web/UC_finding_keys.md [hunk 17]
```diff
@@ -0,0 +1,20 @@
...
+  contents of the search field.
```
Comment (markus2330) [inline doc/usecases/elektra_web/UC_finding_keys.md]: No description what happens if (multiple) key(s) is/are found. Will it/they be shown in the tree?

File: doc/usecases/elektra_web/UC_setup.md [hunk 18]
```diff
@@ -0,0 +1,21 @@
...
+
+## Summary
+
+Title: Setup
+Scope: Setup
+Level: User Goal
+Actors: User (usually a sysadmin)
+Brief: User sets up elektra-web according to the corresponding README.
```
Comment (markus2330) [inline doc/usecases/elektra_web/UC_setup.md]: Why not a script? Too detailed for a use case.

File: doc/usecases/elektra_web/UC_setup_instance.md [hunk 19]
```diff
@@ -0,0 +1,25 @@
...
+Level: User Goal
+Actors: User (usually a sysadmin)
+Brief: User sets up a daemon on the instance and connects it to elektra-web.
+
+## Scenarios
+
+Precondition: Setup
+Main success scenario: User successfully sets up elektrad on the instance (by
+	installing the dependencies and starting elektrad) and connects it to
+	elektra-web (clusterd) by specifying the host and an API key in a config file.
```
Comment (markus2330) [inline doc/usecases/elektra_web/UC_setup_instance.md]: rename clusterd?

File: doc/usecases/elektra_web/UC_duplicating_keys.md (hunk 14 above)
Comment (omnidan) [inline doc/usecases/elektra_web/UC_duplicating_keys.md]: @markus2330 I was going to add an icon to duplicate keys, because this will be a pretty common use case (e.g. creating a new student with the same fields)

File: doc/usecases/elektra_web/UC_setup_instance.md (hunk 19 above)
Comment (omnidan) [inline doc/usecases/elektra_web/UC_setup_instance.md]: @markus2330 I have already implemented cluster configuration in the backend, just decided to remove it in the frontend to keep the focus on the editor. I would keep the name because it still provides that functionality, but if you have a better idea I can also rename it :smile:

File: doc/usecases/elektra_web/UC_finding_keys.md (hunk 17 above)
Comment (omnidan) [inline doc/usecases/elektra_web/UC_finding_keys.md]: As described, the search field works like a filter, so all keys that match the search input will be shown, the other keys will be hidden. I described this more clearly now.

**Comment by omnidan** [comment]
@markus2330 should be fine now!

**Comment by omnidan** [comment]
I just realized that the markdown renderer does not put line breaks unless you make two in the file. This is also a problem with the other use cases, e.g. https://github.com/ElektraInitiative/libelektra/blob/master/doc/usecases/snippet_sharing/UC_convert_snippet.md

We could either append `<br>` after the lines, or use two line breaks instead of one.

File: doc/usecases/elektra_web/UC_undo_redo.md [hunk 20]
```diff
@@ -0,0 +1,20 @@
...
+Scope: Configuration
+Level: User Goal
+Actors: User (usually a sysadmin)
+Brief: User undos/redos changes to a configuration.
+
+## Scenarios
+
+Precondition: Modifying keys, Adding keys, Duplicating keys, Drag & Drop keys
+Main success scenario: User successfully undos/redos a configuration change.
+Alternative scenario: Instance not online. The user is informed about the issue.
```
Comment (markus2330) [inline doc/usecases/elektra_web/UC_undo_redo.md]: Seems to be a copy&paste error from earlier versions? Instances going offline would be an error scenario?

File: doc/usecases/elektra_web/UC_setup_instance.md [hunk 21]
```diff
@@ -0,0 +1,25 @@
...
+
+Title: Setup instance
+Scope: Setup
+Level: User Goal
+Actors: User (usually a sysadmin)
+Brief: User sets up a daemon on the instance and connects it to elektra-web.
+
+## Scenarios
+
+Precondition: elektra-web is set up
```
Comment (markus2330) [inline doc/usecases/elektra_web/UC_setup_instance.md]: We are already finished before starting?

File: doc/usecases/elektra_web/UC_setup_instance.md [hunk 22]
```diff
@@ -0,0 +1,25 @@
...
+Main success scenario: User successfully sets up elektrad on the instance (by
+	installing the dependencies and starting elektrad) and connects it to
+	elektra-web (clusterd) by specifying the host and an API key in a config file.
```
Comment (markus2330) [inline doc/usecases/elektra_web/UC_setup_instance.md]: Better to drop the name `clusterd` and simply call it elektra-web everywhere.

File: doc/usecases/elektra_web/UC_setup_instance.md [hunk 23]
```diff
@@ -0,0 +1,25 @@
...
+Title: Setup instance
+Scope: Setup
+Level: User Goal
+Actors: User (usually a sysadmin)
+Brief: User sets up a daemon on the instance and connects it to elektra-web.
+
+## Scenarios
+
+Precondition: elektra-web is set up
+Main success scenario: User successfully sets up elektrad on the instance (by
```
Comment (markus2330) [inline doc/usecases/elektra_web/UC_setup_instance.md]: As already said earlier: too many details, keep the use cases high-level. Remove whats written in () and in the following sentence (API key, ...).

File: doc/usecases/elektra_web/UC_modifying_keys.md [hunk 24]
```diff
@@ -0,0 +1,25 @@
...
+Level: User Goal
+Actors: User (usually a sysadmin)
+Brief: User modifies a key's contents.
+
+## Scenarios
+
+Precondition: View configuration of an instance
+Main success scenario: User edits a key's contents in a field that is
+  specifically formatted according to its metadata (e.g. number/date fields vs
+	string fields with regex validation). Arrays are specifically formatted as
```
Comment (markus2330) [inline doc/usecases/elektra_web/UC_modifying_keys.md]: tab space mixture?

File: doc/usecases/elektra_web/UC_modifying_keys.md [hunk 25]
```diff
@@ -0,0 +1,25 @@
...
+	lists of fields under a key. A description of the key is also shown.
+Alternative scenario: User enters data that fails the field validation, the
+  input field is marked as invalid (e.g. with a red color and some information
+  text) and the data is NOT sent to the backend.
```
Comment (markus2330) [inline doc/usecases/elektra_web/UC_modifying_keys.md]: Use markdown formatting instead of all-caps?

File: doc/usecases/elektra_web/UC_modifying_keys.md [hunk 26]
```diff
@@ -0,0 +1,25 @@
...
+Error scenario: Technical problems while persisting to the key database. The
+  user is informed about the issue.
```
Comment (markus2330) [inline doc/usecases/elektra_web/UC_modifying_keys.md]: The word "problem" seems to fit better than "issue" here.

**Comment by omnidan** [comment]
@markus2330 sorry it took so long to rework again - I fixed the formatting issues and reworked the texts again.

I hope everything is fine now :grin:

File: doc/usecases/elektra_web/UC_duplicating_keys.md (hunk 14 above)
Comment (markus2330) [inline doc/usecases/elektra_web/UC_duplicating_keys.md]: I somehow doubt that it is a common operation outside of the student use case (which is not relevant anymore). I rather think that common use cases are:

- add array member (`#N+1`)
- add a key within a specific namespace for which the specification is given (but no key exists)

File: doc/usecases/elektra_web/README.md [hunk 27]
```diff
@@ -0,0 +1,17 @@
...
+
+Now we are in the main part of the application, and we can do the following use cases:
+
+- [Adding keys](./UC_adding_keys.md)
+- [Modifying keys](./UC_modifying_keys.md)
+- [Duplicating keys](./UC_duplicating_keys.md)
+- [Drag & Drop keys](./UC_drag_n_drop.md)
+- [Finding keys](./UC_finding_keys.md)
+
+Any of these use cases (except for finding because no change is made), are extended by the [Key validation](./UC_key_validation.md) use case. After doing any of these use cases, we can do the [Undo/Redo](./UC_undo_redo.md) use case to revert them.
```
Comment (markus2330) [inline doc/usecases/elektra_web/README.md]: Good intro!

File: doc/usecases/elektra_web/UC_adding_keys.md [hunk 28]
```diff
@@ -0,0 +1,21 @@
+# Use Case: Adding keys
+
+## Summary
+
+Title: Adding keys
```
Comment (markus2330) [inline doc/usecases/elektra_web/UC_adding_keys.md]: Formatting still broken. An itemization with `-` might be a good solution.

File: doc/usecases/elektra_web/UC_adding_keys.md [hunk 29]
```diff
@@ -0,0 +1,21 @@
...
+Scope: Configuration
+Level: User Goal
+Actors: User (usually a sysadmin)
+Brief: User adds a key (to a subtree).
+
+## Scenarios
+
+Precondition: View configuration of an instance
```
Comment (markus2330) [inline doc/usecases/elektra_web/UC_adding_keys.md]: Sentences should have an dot `.` at the end.

File: doc/usecases/elektra_web/UC_finding_keys.md [hunk 30]
```diff
@@ -0,0 +1,24 @@
...
+Actors: User (usually a sysadmin)
+Brief: User searches for a key by filtering the key database.
+
+## Scenarios
+
+Precondition: View configuration of an instance
+Main success scenario: User searches for a key by entering information (its path
+  or value) in a field. The whole key database is filtered according to the
+  contents of the search field. At first, only keys that match the path are
+  shown. Then, the application will continue searching for values that contain
```
Comment (markus2330) [inline doc/usecases/elektra_web/UC_finding_keys.md]: enumeration/itemize would make it easier to read

File: doc/usecases/elektra_web/UC_modifying_keys.md [hunk 31]
```diff
@@ -0,0 +1,25 @@
...
+  string fields with regex validation). Arrays are specifically formatted as
+  lists of fields under a key. A description of the key is also shown.
+Alternative scenario: User enters data that fails the field validation, the
+  input field is marked as invalid (e.g. with a red color and some information
+  text) and the data is *not* sent to the backend.
+Error scenario: Technical problems while persisting to the key database. The
+  user is informed about the problem.
+Postcondition: The updated key is persisted to the database.
+Non-functional Constraints:
+  - Essential functionality
```
Comment (markus2330) [inline doc/usecases/elektra_web/UC_modifying_keys.md]: is this a non-functional constraint?

File: doc/usecases/elektra_web/UC_setup_instance.md [hunk 32]
```diff
@@ -0,0 +1,23 @@
...
+  connects it to elektra-web.
+Alternative scenario: Authentication issues when connecting the services,
+  e.g. wrong API key.
+Alternative scenario: The instance is already connected to elektra-web. A
+  message informs the user about this.
+Error scenario: Technical problems - the user is informed about the problem.
+Postcondition: User can now access the instance via the client.
+Non-functional Constraints:
+  - Security mechanism
+  - Is precondition to use configuration functions
```
Comment (markus2330) [inline doc/usecases/elektra_web/UC_setup_instance.md]: is this a non-functional constraint? (or a precondition of the others)

**Comment by omnidan** [comment]
I fixed the issues mentioned in the comments and reformatted all usecases. See #1550 for the PR with the updated template file.

**Comment by markus2330** [comment]
Looks much better! Please add "ready to merge" if ready to merge.

**Comment by omnidan** [comment]
@markus2330 done! :ok_hand:
